packages/ember-application/lib/system/application.js

```diff
@@ -456,50 +577,74 @@ var Application = Namespace.extend(RegistryProxy, {
+    object when the boot process is complete.
 
-    @private
-    @deprecated
-    @method initialize
-   **/
-  initialize() {
-    deprecate('Calling initialize manually is not supported. Please see Ember.Application#advanceReadiness and Ember.Application#deferReadiness');
-  },
+    Run any application initializers and run the application load hook. These hooks may
+    choose to defer readiness. For example, an authentication hook might want to defer
+    readiness until the auth token has been retrieved.
 
-  /**
-    Initialize the application. This happens automatically.
+    By default, this method is called automatically on "DOM ready"; however, if autoboot
+    is disabled, you would need to call this method yourself. Calling this method before
+    "DOM ready" is not recommended.
 
-    Run any initializers and run the application load hook. These hooks may
-    choose to defer readiness. For example, an authentication hook might want
-    to defer readiness until the auth token has been retrieved.
-
-    @private
-    @method domReady
+    @public
+    @method boot
+    @return {Promise<Ember.Application,Error>}
   */
-  domReady() {
-    if (this.isDestroyed) { return; }
+  boot() {
+    if (this._bootPromise) { return this._bootPromise; }
 
-    this.boot();
+    try {
+      this._bootSync();
+    } catch(_) {
+      // Ignore th error: in the asynchronous boot path, the error is already reflected
+      // in the promise rejection
+    }
 
-    return this;
+    return this._bootPromise;
   },
 
-  boot() {
-    if (this._bootPromise) { return this._bootPromise; }
+  /**
+    @private
+  */
+  _booted: false,
 
-    var defer = new Ember.RSVP.defer();
-    this._bootPromise = defer.promise;
-    this._bootResolver = defer;
+  /**
+    Unfortunately, a lot of existing code assumes the booting process is
+    "synchronous". Specifically, a lot of tests assumes the last call to
+    `app.advanceReadiness()` or `app.reset()` will result in the app being
+    fully-booted when the current runloop completes.
 
-    this.runInitializers();
-    runLoadHooks('application', this);
+    We would like new code (like the `visit` API) to stop making this assumption,
+    so we created the asynchronous version above that returns a promise. But until
+    we have migrated all the code, we would have to expose this method for use
+    *internall* in places where we need to boot an app "synchronously".
 
-    this.advanceReadiness();
+    @private
+  */
+  _bootSync() {
+    if (this._booted) { return; }
+
+    // Even though this returns synchronously, we still need to make sure the
+    // boot promise exists for book-keeping purposes: if anything went wrong in
+    // the boot process, we need to store the error as a rejection on the boot
+    // promise so that a future caller of `boot()` can tell what failed.
+    let defer = this._bootResolver = new Ember.RSVP.defer();
```

Comment:
This is the kind of thing I want people to not accidentally do re above :P (if I were you I would be tempted to think this is ugly and do that refactor too).

We actually cannot :( (I also didn't introduce this, `_bootPromise` and `_bootResolver` were from original patch, they just didn't correctly reject on errors.) The main reason is that this is really a function that's broken into two halves (the work it's doing here, plus the continuation at `didBecomeReady`). If you follow that code you will see that the fulfillment is in the other half, so we actually need to save of the `resolve`/`reject`somewhere.

(And unfortunately, not using the promise constructor forces us to try/catch manually, which is the bug I'm trying to fix vs the implementation on master.)